### PR TITLE
MockTransportService should physically disconnect when simulating it

### PR DESCRIPTION
**Pull request by bleskes**
This is in order to trigger listeners for disconnect events, most importantly the NodeFaultDetection

This is the cause of failure in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.2+multijob-unix-compatibility/os=debian/72

### CODE REVIEW

**Comment by bleskes** [comment]
please hold with reviews. CI runs have show some interesting failures. I'm still working on this.

**Comment by bleskes** [comment]
I had this PR a step further. MockTransportService now does slightly better job at mimicking real life failures: connecting to already connected node will be a noop (we don't detect any errors here in production either) and failing to send will cause the target node to be disconnected.

The reason for the above is interesting as well - in my previous iteration, a connectToNode call on an already connected node will disconnect the node. This changed dymanics a bit and caused the standard logic in NodeConnectionService to disconnect nodes aggressively (as it calls connectToNode on every CS update). This in turn caused a shard started action, sent from IndicesClusterStateService, to hit a node not connected exception. The later meant it went into a retry via an observer which caused an assertion to trigger that the CS isn't sampled from the CS update thread. I'm not sure we want to forbid a retry loop at that level. I will discuss this with @ywelsch as a potential follow up.

**Review by s1monw — APPROVED** [review]
LGTM

File: test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java [hunk 1]
```diff
@@ -421,6 +431,37 @@ private LookupTestTransport transport() {
...
+        if (transport.nodeConnected(node)) {
+            // this a connected node, disconnecting from it will be up the exception
+            transport.disconnectFromNode(node);
+        } else {
+            throw new ConnectTransportException(node, reason, e);
+        }
+    }
+
+    /**
+     * simulates a disconnect by closing the connection and and throwing a
```
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java]: and and

File: test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java [hunk 2]
```diff
@@ -421,6 +431,37 @@ private LookupTestTransport transport() {
     }
 
     /**
+     * simulates a disconnect by disconnecting from the underlying transport and throwing a
+     * {@link ConnectTransportException}
+     */
+    private void simulateDisconnect(DiscoveryNode node,Transport transport, String reason) {
```
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java]: space missing

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**